### PR TITLE
JAMES-2583 New JoiningBlobStore impl

### PR DESCRIPTION
**Pull request by trantienduchn**
- based on `BlobStore` interface
- compose one `BlobStore` as primary and one `BlobStore` as secondary
- read from primary, if fails, null result, emtpy result => fallback to secondary
- write to primary, if fails, null result, emtpy result => fallback to secondary

### CODE REVIEW

File: server/blob/blob-joining/src/main/java/org/apache/james/blob/joining/JoiningBlobStore.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+    @Override
+    public CompletableFuture<BlobId> save(byte[] data) {
+        try {
+            return saveToPrimaryFallbackIfFails(
+                primaryBlobStore.save(data),
+                () -> secondaryBlobStore.save(data));
+        } catch (Exception e) {
+            LOGGER.error("exception directly happens while saving bytes data, fall back to secondary blob store", e);
+            return secondaryBlobStore.save(data);
```
Comment (aduprat) [inline server/blob/blob-joining/src/main/java/org/apache/james/blob/joining/JoiningBlobStore.java]: I think this operation has already been done, no?
Then you should throw here.

File: server/blob/blob-joining/src/main/java/org/apache/james/blob/joining/JoiningBlobStore.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+    @Override
+    public CompletableFuture<BlobId> save(InputStream data) {
+        try {
+            return saveToPrimaryFallbackIfFails(
+                primaryBlobStore.save(data),
+                () -> secondaryBlobStore.save(data));
+        } catch (Exception e) {
+            LOGGER.error("exception directly happens while saving InputStream data, fall back to secondary blob store", e);
+            return secondaryBlobStore.save(data);
```
Comment (aduprat) [inline server/blob/blob-joining/src/main/java/org/apache/james/blob/joining/JoiningBlobStore.java]: same (also in other _save_ & _read_ public methods)

File: server/blob/blob-joining/src/main/java/org/apache/james/blob/joining/JoiningBlobStore.java (hunk 1 above)
Comment (chibenwa) [reply server/blob/blob-joining/src/main/java/org/apache/james/blob/joining/JoiningBlobStore.java]: No not necessarlily, it depends at which point you fail.

A failure can be **an exception beuing thrown** or **an exception during the future execution**. Both needs to be addressed.

File: server/blob/blob-joining/src/main/java/org/apache/james/blob/joining/JoiningBlobStore.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    @Override
+    public CompletableFuture<byte[]> readBytes(BlobId blobId) {
+        return primaryBlobStore.readBytes(blobId)
+            .thenCompose(primaryResultInBytes -> readFromSecondaryIfNeeded(blobId, primaryResultInBytes));
+    }
+
+    @Override
+    public InputStream read(BlobId blobId) {
+        return Optional.ofNullable(primaryBlobStore.read(blobId))
+            .filter(Throwing.<InputStream>predicate(inputStream -> inputStream.available() > 0).sneakyThrow())
```
Comment (rouazana) [inline server/blob/blob-joining/src/main/java/org/apache/james/blob/joining/JoiningBlobStore.java]: are you sure you can rely on inputStream.available()? I was pretty sure it's not very reliable. See https://docs.oracle.com/javase/7/docs/api/java/io/InputStream.html#available()

File: server/blob/blob-joining/src/main/java/org/apache/james/blob/joining/JoiningBlobStore.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            .orElseGet(() -> secondaryBlobStore.read(blobId));
+    }
+
+    private CompletionStage<byte[]> readFromSecondaryIfNeeded(BlobId blobId, byte[] primaryResultInBytes) {
+        if (isNullOrEmpty(primaryResultInBytes)) {
```
Comment (rouazana) [inline server/blob/blob-joining/src/main/java/org/apache/james/blob/joining/JoiningBlobStore.java]: maybe our blob api is not very good if we need to check the content is empty or not... but that's for an other ticker/pr

File: server/blob/blob-joining/src/test/java/org/apache/james/blob/joining/JoiningBlobStoreTest.java [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        }
+    }
+
+    private static final HashBlobId.Factory BLOB_ID_FACTORY = new HashBlobId.Factory();
+    private static final MemoryBlobStore PRIMARY_BLOB_STORE = new MemoryBlobStore(BLOB_ID_FACTORY);
+    private static final MemoryBlobStore SECONDARY_BLOB_STORE = new MemoryBlobStore(BLOB_ID_FACTORY);
+    private static final ThrowingBlobStore THROWING_BLOB_STORE = new ThrowingBlobStore();
+
+    private static final byte [] BLOB_CONTENT = "blob content".getBytes();
+    private static final byte [] BLOB_CONTENT_CAUSES_THROWING_DIRECTLY = "blob content causes throwing directly".getBytes();
```
Comment (rouazana) [inline server/blob/blob-joining/src/test/java/org/apache/james/blob/joining/JoiningBlobStoreTest.java]: these constants could be before the ThrowingBlobStore, no? (or even inside?)

File: server/blob/blob-joining/src/main/java/org/apache/james/blob/joining/JoiningBlobStore.java (hunk 1 above)
Comment (trantienduchn) [reply server/blob/blob-joining/src/main/java/org/apache/james/blob/joining/JoiningBlobStore.java]: yeah, `secondaryBlobStore.save(data)` in try block is passed as a supplier. I don't execute it util first completableFuture completed exceptionally. This one is for addressing execption happens directly in method call, not in while executing future.

File: server/blob/blob-joining/src/test/java/org/apache/james/blob/joining/JoiningBlobStoreTest.java (hunk 5 above)
Comment (chibenwa) [reply server/blob/blob-joining/src/test/java/org/apache/james/blob/joining/JoiningBlobStoreTest.java]: Do we still need `BLOB_CONTENT_CAUSES_THROWING_DIRECTLY` btw?

File: server/blob/blob-joining/src/main/java/org/apache/james/blob/joining/JoiningBlobStore.java (hunk 3 above)
Comment (trantienduchn) [reply server/blob/blob-joining/src/main/java/org/apache/james/blob/joining/JoiningBlobStore.java]: oh, I didn't read it carefully. I will find another way

**Comment by trantienduchn** [comment]
> About 95655a6 (completely fallback to secondary) I was not aware that we wanted this feature. Are you sure?

I've just think about it this morning. And I want people's ideas

File: server/blob/blob-joining/src/test/java/org/apache/james/blob/joining/JoiningBlobStoreTest.java (hunk 5 above)
Comment (trantienduchn) [reply server/blob/blob-joining/src/test/java/org/apache/james/blob/joining/JoiningBlobStoreTest.java]: This change is old. I remove it in the second commit 95655a6 (completely fallback to secondary)

**Comment by aduprat** [comment]
> I've just think about it this morning. And I want people's ideas

According to a comment in the issue, you should save in only one blobStore:
```
You need to explain what that JoiningObjectStore would be.
Build on top of a primary and a secondary blobstore, this blob store will...

* Read from primary  then if needed in secondary
* Write to secondary
```

File: server/blob/blob-joining/pom.xml [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        <artifactId>james-server-blob</artifactId>
+        <groupId>org.apache.james</groupId>
+        <version>3.2.0-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <artifactId>blob-joining</artifactId>
+    <packaging>jar</packaging>
+
+    <name>Apache James :: Server :: Blob :: Joining Object storage</name>
```
Comment (mbaechler) [inline server/blob/blob-joining/pom.xml]: you should describe "Joining" here

File: server/blob/blob-joining/src/main/java/org/apache/james/blob/joining/JoiningBlobStore.java [hunk 7]
```diff
@@ -94,13 +94,10 @@ public InputStream read(BlobId blobId) {
...
     private Optional<InputStream> readFromPrimary(BlobId blobId) throws IOException {
-        InputStream inputStream = primaryBlobStore.read(blobId);
-        byte[] bytes = IOUtils.toByteArray(inputStream);
-        if (hasContent(bytes)) {
-            return Optional.of(new ByteArrayInputStream(bytes));
-        }
-
-        return Optional.empty();
+        return Optional.ofNullable(primaryBlobStore.read(blobId))
+            .map(Throwing.<InputStream, byte[]>function(IOUtils::toByteArray).sneakyThrow())
```
Comment (rouazana) [inline server/blob/blob-joining/src/main/java/org/apache/james/blob/joining/JoiningBlobStore.java]: if we transform the inputStream in byte array, we loose all the benefits of the InputStream, no?

Comment (trantienduchn) [reply server/blob/blob-joining/src/main/java/org/apache/james/blob/joining/JoiningBlobStore.java]: Yeah, but If I don't transform it to byte arrays, there is no way to determine whether that inpustream is empty or not.

**Comment by trantienduchn** [comment]
```
[852f0585e471809755f431de4461ea4130665ed7] [ERROR] Error occurred in starting fork, check output in log
[852f0585e471809755f431de4461ea4130665ed7] [ERROR] Process Exit Code: 100
[852f0585e471809755f431de4461ea4130665ed7] [ERROR] Crashed tests:
[852f0585e471809755f431de4461ea4130665ed7] [ERROR] org.apache.james.webadmin.routes.SieveQuotaRoutesTest
[852f0585e471809755f431de4461ea4130665ed7] [ERROR] org.apache.maven.surefire.booter.SurefireBooterForkException: ExecutionException The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
[852f0585e471809755f431de4461ea4130665ed7] [ERROR] Command was /bin/sh -c cd /james-project/server/protocols/webadmin/webadmin-data && /usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java -Xms512m -Xmx1024m -jar /james-project/server/protocols/webadmin/webadmin-data/target/surefire/surefirebooter421573488808001669.jar /james-project/server/protocols/webadmin/webadmin-data/target/surefire 2018-11-06T03-27-25_553-jvmRun5 surefire224388175850545900tmp surefire_3186558052992624210885tmp
```
test this please

**Comment by trantienduchn** [comment]
I added fallback completely commit. And Renaming it from `Joining` to `Union` and variables also. I didn't rebase to change the names because it is a nightmare, full of conflicts.

Renaming in the last commit.

**Comment by trantienduchn** [comment]
test this please

File: server/blob/blob-joining/src/main/java/org/apache/james/blob/joining/JoiningBlobStore.java [hunk 8]
```diff
@@ -36,6 +37,7 @@
 public class JoiningBlobStore implements BlobStore {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(JoiningBlobStore.class);
+    private static final int UN_AVAILABLE = -1;
```
Comment (rouazana) [inline server/blob/blob-joining/src/main/java/org/apache/james/blob/joining/JoiningBlobStore.java]: s/UN_AVAILABLE/UNAVAILABLE

File: server/blob/blob-joining/src/main/java/org/apache/james/blob/joining/JoiningBlobStore.java [hunk 9]
```diff
@@ -78,12 +80,22 @@ public InputStream read(BlobId blobId) {
...
+            .filter(Throwing.predicate(this::streamHasContent).sneakyThrow())
+            .<InputStream>map(Function.identity())
             .orElseGet(() -> secondaryBlobStore.read(blobId));
     }
 
+    @VisibleForTesting
+    boolean streamHasContent(PushbackInputStream pushBackIS) throws IOException {
+        int byteRead = pushBackIS.read();
+        if (byteRead != UN_AVAILABLE) {
+            pushBackIS.unread(byteRead);
```
Comment (rouazana) [inline server/blob/blob-joining/src/main/java/org/apache/james/blob/joining/JoiningBlobStore.java]: nice hack!

File: server/blob/blob-union/src/test/java/org/apache/james/blob/union/UnionBlobStoreTest.java [hunk 10]
```diff
@@ -0,0 +1,411 @@
...
+            MemoryBlobStore legacyBlobStore = new MemoryBlobStore(BLOB_ID_FACTORY);
+            UnionBlobStore unionBlobStore = new UnionBlobStore(new FutureThrowingBlobStore(), legacyBlobStore);
+            BlobId blobId = legacyBlobStore.save(BLOB_CONTENT).get();
+
+            assertThat(unionBlobStore.readBytes(blobId).get())
+                .isEqualTo(BLOB_CONTENT);
+        }
+    }
+
+    @TestInstance(TestInstance.Lifecycle.PER_CLASS)
```
Comment (trantienduchn) [inline server/blob/blob-union/src/test/java/org/apache/james/blob/union/UnionBlobStoreTest.java]: normally `@MethodSource` - parameter factory, only able to access static method. I need to declare `CurrentAndLegacyCouldNotComplete` test instance as `PER_CLASS`. This should help `@MethodSource` access to non static method factory. see [junit 5 documentaion](https://junit.org/junit5/docs/5.2.0/user-guide/#writing-tests-parameterized-tests-sources-MethodSource)

File: server/blob/blob-union/src/test/java/org/apache/james/blob/union/UnionBlobStoreTest.java [hunk 11]
```diff
@@ -224,6 +247,61 @@ void readBytesShouldReturnFallbackToLegacyWhenCurrentCompletedExceptionally() th
         }
     }
 
+    @TestInstance(TestInstance.Lifecycle.PER_CLASS)
```
Comment (trantienduchn) [inline server/blob/blob-union/src/test/java/org/apache/james/blob/union/UnionBlobStoreTest.java]: normally `@MethodSource` - parameter factory, only able to access static method. I need to declare `CurrentAndLegacyCouldNotComplete` test instance as `PER_CLASS`. This should help `@MethodSource` access to non static method factory. see [junit 5 documentaion](https://junit.org/junit5/docs/5.2.0/user-guide/#writing-tests-parameterized-tests-sources-MethodSource)

**Comment by trantienduchn** [comment]
test this please

**Comment by trantienduchn** [comment]
```
[46ef8671bde26632c8e7d57155b18129a5a6759c] [ERROR] Errors: 
[46ef8671bde26632c8e7d57155b18129a5a6759c] [ERROR]   MailQueueContract.concurrentEnqueueDequeueShouldNotFail:387 » NotTerminated
[46ef8671bde26632c8e7d57155b18129a5a6759c] [INFO] 
[46ef8671bde26632c8e7d57155b18129a5a6759c] [ERROR] Tests run: 179, Failures: 0, Errors: 1, Skipped: 0
```
test this please

**Pull request merged**